controllers/routemonitor/adder/adder.go

```diff
@@ -79,12 +79,94 @@ func (r *RouteMonitorAdder) EnsureServiceMonitorResourceExists(ctx context.Conte
 	}
 
 	//Update status with serviceMonitorRef
-	routeMonitor.Status.ServiceMonitorRef.Namespace = namespacedName.Namespace
-	routeMonitor.Status.ServiceMonitorRef.Name = namespacedName.Name
-	err := r.Status().Update(ctx, &routeMonitor)
-	if err != nil {
-		return utilreconcile.RequeueReconcileWith(err)
+	desiredRef := v1alpha1.NamespacedName{
+		Name:      namespacedName.Name,
+		Namespace: namespacedName.Namespace,
 	}
+	emptyRef := v1alpha1.NamespacedName{}
 
+	currentRef := routeMonitor.Status.ServiceMonitorRef
+	if currentRef != emptyRef && desiredRef != currentRef {
+		return utilreconcile.RequeueReconcileWith(customerrors.InvalidReferenceUpdate)
+	}
+
+	if currentRef == emptyRef && desiredRef != emptyRef {
+		routeMonitor.Status.ServiceMonitorRef = desiredRef
+
+		err := r.Status().Update(ctx, &routeMonitor)
+		if err != nil {
+			return utilreconcile.RequeueReconcileWith(err)
+		}
+		return utilreconcile.StopReconcile()
+
+	}
+	return utilreconcile.ContinueReconcile()
+}
+func (r *RouteMonitorAdder) EnsurePrometheusRuleResourceExists(ctx context.Context, routeMonitor v1alpha1.RouteMonitor) (utilreconcile.Result, error) {
+	// Was the RouteURL populated by a previous step?
+	if routeMonitor.Status.RouteURL == "" {
+		return utilreconcile.RequeueReconcileWith(customerrors.NoHost)
+	}
+
+	// Is the SloSpec configured on this CR?
+	emptySlo := v1alpha1.SloSpec{}
+	if routeMonitor.Spec.Slo == emptySlo {
+		return utilreconcile.ContinueReconcile()
+	}
+
+	if !routeMonitor.Spec.Slo.IsValid() {
+		return utilreconcile.RequeueReconcileWith(customerrors.InvalidSLO)
+	}
+
+	if !finalizer.HasFinalizer(&routeMonitor, consts.FinalizerKey) {
+		// If the routeMonitor doesn't have a finalizer, add it
+		utilfinalizer.Add(&routeMonitor, routemonitorconst.FinalizerKey)
+		if err := r.Update(ctx, &routeMonitor); err != nil {
+			return utilreconcile.RequeueReconcileWith(err)
+		}
+		return utilreconcile.StopReconcile()
+	}
+
+	namespacedName := types.NamespacedName{Name: routeMonitor.Name, Namespace: routeMonitor.Namespace}
+	normalizedPercent, succeed := routeMonitor.Spec.Slo.NormalizeValue()
+	if !succeed {
+		return utilreconcile.RequeueReconcileWith(customerrors.InvalidSLO)
+	}
+
+	resource := &monitoringv1.PrometheusRule{}
+	populationFunc := func() monitoringv1.PrometheusRule {
```

Comment:
still don't understand this pattern, one day you need to explain it to me so I understand its benefit ;)

Reply:
sure :)